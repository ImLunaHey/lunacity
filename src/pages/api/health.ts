import type { NextApiRequest, NextApiResponse } from 'next'

const Health = (_request: NextApiRequest, response: NextApiResponse) => {
    response.status(200);
};

export default Health;
