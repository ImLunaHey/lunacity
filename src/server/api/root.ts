import { createTRPCRouter } from './trpc';
import { pageRouter } from './routers/page';
import { tagRouter } from './routers/tag';

/**
 * This is the primary router for your server.
 *
 * All routers added in /api/routers should be manually added here
 */
export const appRouter = createTRPCRouter({
  page: pageRouter,
  tag: tagRouter,
});

// export type definition of API
export type AppRouter = typeof appRouter;
