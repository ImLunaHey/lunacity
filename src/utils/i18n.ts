import i18nNext, { type InitOptions } from 'i18next';
import { initReactI18next } from 'react-i18next';

import Backend, { type HttpBackendOptions } from 'i18next-http-backend';
import LanguageDetector from 'i18next-browser-languagedetector';
import getConfig from 'next/config';
import Pseudo from 'i18next-pseudo';
import { env } from '@app/env.mjs';

const { publicRuntimeConfig } = getConfig() as { publicRuntimeConfig: { APP_URL: string; WS_URL: string; } };
const { APP_URL } = publicRuntimeConfig;

const options = {
    fallbackLng: 'en',
    supportedLngs: ['en', 'ar', 'cn', 'es', 'pt'],
    debug: !!env.NEXT_PUBLIC_LANG_DEBUG,
    defaultNS: 'common',
    returnNull: false,
    load: 'languageOnly' as const,
    react: {
        useSuspense: false,
    },
    ns: ['common'],
    interpolation: {
        escapeValue: false, // not needed for react as it escapes by default
    },
    nonExplicitSupportedLngs: true,
    backend: {
        loadPath: `${APP_URL}/locales/{{lng}}/{{ns}}.json`,
    },
    detection: {
        caches: []
    },
    postProcess: ['pseudo'],
} satisfies InitOptions<HttpBackendOptions>;

export const i18n = i18nNext
    .use(new Pseudo({
        enabled: env.NEXT_PUBLIC_PSUDO_ENABLED === 'true',
        languageToPseudo: 'en-US',
    }))
    // load translation using http -> see /public/locales (i.e. https://github.com/i18next/react-i18next/tree/master/example/react/public/locales)
    // learn more: https://github.com/i18next/i18next-http-backend
    .use(Backend)
    // detect user language
    // learn more: https://github.com/i18next/i18next-browser-languageDetector
    .use(LanguageDetector)
    // pass the i18n instance to react-i18next.
    .use(initReactI18next)
    // init i18next
    // for all options read: https://www.i18next.com/overview/configuration-options
    .init<HttpBackendOptions>(options);
