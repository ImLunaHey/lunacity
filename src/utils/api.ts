/**
 * This is the client-side entrypoint for your tRPC API.
 * It's used to create the `api` object which contains the Next.js App-wrapper
 * as well as your typesafe react-query hooks.
 *
 * We also create a few inference helpers for input and output types
 */
import { createWSClient, httpBatchLink, loggerLink, splitLink, wsLink } from '@trpc/client';
import { createTRPCNext } from '@trpc/next';
import { type inferRouterInputs, type inferRouterOutputs } from '@trpc/server';
import { type NextPageContext } from 'next';
import getConfig from 'next/config';
import superJSON from 'superjson';
import { type AppRouter } from '@app/server/api/root';

const { publicRuntimeConfig } = getConfig() as { publicRuntimeConfig: { APP_URL: string; WS_URL: string; } };
const { APP_URL, WS_URL } = publicRuntimeConfig;
const getEndingLink = (ctx: NextPageContext | undefined) => {
  if (typeof window === 'undefined') {
    return httpBatchLink({
      url: `${APP_URL}/api/trpc`,
      headers() {
        if (ctx?.req) {
          // on ssr, forward client's headers to the server
          return {
            ...ctx.req.headers,
            'x-ssr': '1',
          };
        }
        return {};
      },
    });
  }

  // Make Websocket client
  const client = createWSClient({
    url: WS_URL,
  });

  // If we're on the client create a split link that sends both queries
  // and mutations over http-batch and all subscriptions over websocket
  return splitLink({
    condition(op) {
      return op.type !== 'subscription';
    },
    // HTTP batching
    true: httpBatchLink({
      url: `${APP_URL}/api/trpc`,
      headers() {
        if (ctx?.req) {
          // on ssr, forward client's headers to the server
          return {
            ...ctx.req.headers,
            'x-ssr': '1',
          };
        }
        return {};
      },
    }),
    // Websocket
    false: wsLink<AppRouter>({
      client,
    })
  });
}

/**
 * A set of typesafe react-query hooks for your tRPC API
 */
export const api = createTRPCNext<AppRouter>({
  config({ ctx }) {
    return {
      /**
       * Transformer used for data de-serialization from the server
       * @see https://trpc.io/docs/data-transformers
       **/
      transformer: superJSON,

      /**
       * Links used to determine request flow from client to server
       * @see https://trpc.io/docs/links
       * */
      links: [
        loggerLink({
          enabled: (opts) =>
            process.env.NODE_ENV === 'development' ||
            (opts.direction === 'down' && opts.result instanceof Error),
        }),
        getEndingLink(ctx)
      ]
    };
  },
  /**
   * Whether tRPC should await queries when server rendering pages
   * @see https://trpc.io/docs/nextjs#ssr-boolean-default-false
   */
  ssr: false,
});

/**
 * Inference helper for inputs
 * @example type HelloInput = RouterInputs['example']['hello']
 **/
export type RouterInputs = inferRouterInputs<AppRouter>;
/**
 * Inference helper for outputs
 * @example type HelloOutput = RouterOutputs['example']['hello']
 **/
export type RouterOutputs = inferRouterOutputs<AppRouter>;
