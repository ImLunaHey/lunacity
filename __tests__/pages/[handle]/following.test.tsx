import '@testing-library/jest-dom';
import { api } from '@app/utils/api';
import Following, { getServerSideProps } from '../../../src/pages/[handle]/following';
import type { IncomingMessage, ServerResponse } from 'http';
import { render } from '@testing-library/react';

jest.mock('@app/common/with-public-access', () => ({
  // eslint-disable-next-line @typescript-eslint/no-unsafe-return, @typescript-eslint/no-unsafe-call, @typescript-eslint/no-explicit-any
  withPublicAccess: (fn: any) => (ctx: any) => fn(ctx),
}));

// Mock next-auth with defaults
jest.mock('next/config', () => ({
  __esModule: true,
  default: () => ({
    publicRuntimeConfig: {
      APP_URL: 'http://localhost:3000',
      WS_URL: 'ws://localhost:3001',
    },
  }),
}));

jest.mock('@app/utils/api', () => ({
  api: {
    page: {
      getPageFollowing: {
        useQuery: jest.fn(() => ({
          isLoading: true,
          data: undefined,
        })),
      },
    },
  },
}));

jest.mock('react-i18next', () => ({
  useTranslation: jest.fn(() => ({
    t: jest.fn((text: string, args) => (args ? `${text} ${JSON.stringify(args, null)}` : text)),
    ready: true,
  })),
}));

describe('getServerSideProps', () => {
  it('returns the handle from the url', async () => {
    const handle = '@staff';
    const result = await getServerSideProps({
      query: {
        handle,
      },
      req: {} as IncomingMessage & { cookies: Partial<{ [key: string]: string }> },
      res: {} as ServerResponse<IncomingMessage>,
      resolvedUrl: '/',
    });
    expect(result).toEqual({
      props: {
        handle: 'staff',
      },
    });
  });

  it('renders the following page', () => {
    // @ts-expect-error - Mocking useQuery
    jest.mocked(api.page.getPageFollowing.useQuery).mockImplementation(() => ({
      isLoading: false,
      data: [],
    }));

    const { container } = render(<Following handle={'staff'} />);
    expect(container).toMatchSnapshot();
  });
});
