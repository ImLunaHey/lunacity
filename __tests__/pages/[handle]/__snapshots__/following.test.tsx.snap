// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getServerSideProps renders the following page 1`] = `<div />`;
